service/tasks/volume.py

```diff
@@ -30,45 +31,33 @@
       default_retry_delay=20,
       ignore_result=False)
 def check_volume_task(driverCls, provider, identity,
-                      instance_id, volume_id, *args, **kwargs):
+                      instance_id, volume_id, device_type='ext4', *args, **kwargs):
     try:
         celery_logger.debug("check_volume task started at %s." % datetime.now())
         driver = get_driver(driverCls, provider, identity)
         instance = driver.get_instance(instance_id)
         volume = driver.get_volume(volume_id)
+        username = identity.get_username()
         attach_data = volume.extra['attachments'][0]
-        device = attach_data['device']
-
-        private_key = ATMOSPHERE_PRIVATE_KEYFILE
-        kwargs.update({'ssh_key': private_key})
-        kwargs.update({'timeout': 120})
-
-        # One script to make two checks:
-        # 1. Voume exists 2. Volume has a filesystem
-        cv_script = check_volume(device)
-        # NOTE: non_zero_deploy needed to stop LibcloudDeploymentError from being
-        # raised
-        kwargs.update({'deploy': cv_script,
-                       'non_zero_deploy': True})
-        driver.deploy_to(instance, **kwargs)
-        kwargs.pop('non_zero_deploy', None)
-        # Script execute
-
-        if cv_script.exit_status != 0:
-            if 'No such file' in cv_script.stdout:
-                raise Exception('Volume check failed: %s. '
-                                'Device %s does not exist on instance %s'
-                                % (volume, device, instance))
-            elif 'Bad magic number' in cv_script.stdout:
-                # Filesystem needs to be created for this device
-                celery_logger.info("Mkfs needed")
-                mkfs_script = mkfs_volume(device)
-                kwargs.update({'deploy': mkfs_script})
-                driver.deploy_to(instance, **kwargs)
-            else:
-                raise Exception('Volume check failed: Something weird')
-
-        celery_logger.debug("check_volume task finished at %s." % datetime.now())
+        device_location = attach_data['device']
+        celery_logger.info("device_location: %s" % device_location)
+
+        # One playbook to make two checks:
+        # 1. Voume exists
+        # 2. Volume has a filesystem
+        #    (If not, create one of type 'device_type')
+        playbooks = deploy_check_volume(
+            instance.ip, username, instance.id,
+            device_location, device_type=device_type)
+        celery_logger.info(playbooks.__dict__)
+        hostname = build_host_name(instance.id, instance.ip)
+        result = False if execution_has_failures(playbooks, hostname)\
```

Comment:
This is confusing to me but perhaps because of my ignorance. Instead of setting `result = False` and raising a generic exception, why wouldn't we raise exceptions close to where they happen (e.g. inside of whatever `execution_has_failures()` does) and catch / re-raise them here, maybe exposing more error detail to the API client?